lfs/batcher.go

```diff
@@ -6,36 +6,43 @@ import "sync/atomic"
 // be added to the batcher from multiple goroutines and pulled off in groups
 // when one of the following conditions occurs:
 //   * The batch size is reached
+//   * Flush() is called, forcing the batch to be returned immediately, as-is
 //   * Exit() is called
-// When an Exit() occurs, the group may be smaller than the batch size.
+// When an Exit() or Flush() occurs, the group may be smaller than the batch
+// size.
 type Batcher struct {
 	exited     uint32
 	batchSize  int
 	input      chan interface{}
 	batchReady chan []interface{}
+	flush      chan interface{}
 }
 
 // NewBatcher creates a Batcher with the batchSize.
 func NewBatcher(batchSize int) *Batcher {
 	b := &Batcher{
 		batchSize:  batchSize,
-		input:      make(chan interface{}, batchSize),
+		input:      make(chan interface{}),
```

Comment:
I think this is an interesting thought. The reason I un-buffered the channel in https://github.com/github/git-lfs/pull/1528/commits/98873d87c4b06beecbba57221e70dd975ce7d7d4 and https://github.com/github/git-lfs/pull/1528/commits/a27d92d1a48f5d6d57218501b60f00ebbbf75826 was so that I could ensure that after calling `Add()` a number of times in test, that they would _for sure_ be in the batch so that after `Flush()`-ing, I'd get them out.

However, I'm not actually sure we ever needed the buffered behavior in the first place. I could be wrong, since I haven't been in this part of the code in a while, but I ran some benchmarks to get a second opinion.

With buffering (on master):
```
~/g/g/lfs (master) $ g rev-parse HEAD
b7b1f2b7985c8f7a18c563c70898f669ffb8c13c

~/g/g/lfs (master!) $ go test -bench=.
BenchmarkLsTreeParser-4           500000              4087 ns/op
PASS
ok      github.com/github/git-lfs/lfs   2.889s
```

Without buffering (on 49489c0f43011c5f0638eb2fd17b993e7488a0a6):
```
~/g/g/lfs (batcher-truncate!) $ g rev-parse HEAD
49489c0f43011c5f0638eb2fd17b993e7488a0a6

~/g/g/lfs (batcher-truncate!) $ go test -bench=.
BenchmarkLsTreeParser-4           500000              4096 ns/op
PASS
ok      github.com/github/git-lfs/lfs   2.877s
```

I ran these benchmarks a number of times locally to minimize any jitter, but the numbers stayed remarkably consistent.

If we were to buffer the channel again, I'm thinking we could either:
- Upon receiving a signal from `<-b.flush`, enter a spin-loop until we have drained the channel (which has the 👎  side of thrashing the user's CPU)
- Increment a counter of "number of times flushed", and check after either a `close(b.input)` or some other signal, and preform the same drain as above.

But, given the added complexity of the ideas I'm thinking about above, and the minimal difference in performance, I think we'd be fine to keep this as-is.

For reference, the benchmark I ran is in [this gist]().